File: src/content/algorithms/coin-change/code.py
def coin_change(coins, amount):  # #@function-def
    INF = float('inf')
    dp = [INF] * (amount + 1)  # #@init-dp
    dp[0] = 0
    choice = [None] * (amount + 1)

    for j in range(1, amount + 1):  # #@outer-loop
        for c in coins:  # #@inner-loop
            if c <= j and dp[j - c] != INF:  # #@validity-check
                candidate = dp[j - c] + 1  # #@recurrence
                if candidate < dp[j]:
                    dp[j] = candidate
                    choice[j] = c

    if dp[amount] == INF:  # #@return-result
        return -1, []

    # Backtrack para reconstruir as moedas
    coins_used = []
    j = amount
    while j > 0:
        coins_used.append(choice[j])
        j -= choice[j]

    return dp[amount], coins_used

File: src/content/algorithms/coin-change/code.rs
pub fn coin_change(coins: &[i32], amount: i32) -> i32 {  // #@generic-function
    let n = amount as usize;
    let mut dp: Vec<Option<i32>> = vec![None; n + 1];  // #@init-dp
    dp[0] = Some(0);

    for j in 1..=n {  // #@outer-loop
        for &c in coins {  // #@inner-loop
            let c = c as usize;
            if c <= j {  // #@validity-check
                if let Some(prev) = dp[j - c] {
                    let candidate = prev + 1;  // #@recurrence
                    dp[j] = Some(match dp[j] {
                        Some(cur) => cur.min(candidate),
                        None => candidate,
                    });
                }
            }
        }
    }

    dp[n].unwrap_or(-1)  // #@return-result
}
